File: packages/shared/src/session.ts
/**
 * Session sidecar contract (`session.json` + `session.journal.jsonl`).
 *
 * The sidecar is the SOURCE OF TRUTH for marks. OBS Hybrid MP4 chapters are a
 * redundant convenience copy: they are written at file finalization and are
 * lost if OBS crashes mid-session (tech-stack report §3.2), so nothing
 * downstream may depend on them. The journal is an append-only crash log the
 * pipeline can rebuild `session.json` from.
 *
 * All times are integer-ish milliseconds. Two clock domains exist:
 *  - `monoMs`  — monotonic milliseconds since an arbitrary process epoch
 *                (wall clocks jump; monotonic clocks don't).
 *  - `videoMs` — position in the OBS recording, anchored via
 *                `GetRecordStatus.outputDuration` (immune to StartRecord
 *                latency and PauseRecord skew; tech-stack report §3.4).
 */

export const SIDECAR_SCHEMA_VERSION = 1 as const;

export const SIDECAR_FILENAME = 'session.json';
export const JOURNAL_FILENAME = 'session.journal.jsonl';

/** What is being recorded. No assumption is made about the target (§3.1). */
export type CaptureTargetKind = 'game' | 'video' | 'app' | 'device' | 'other';

export interface CaptureTarget {
  kind: CaptureTargetKind;
  /** Free-form display name, e.g. "DungeonSweeper2" or "scrcpy mirror". */
  name?: string;
}

export type MarkKind = 'manual' | 'vad' | 'auto';

/** How a mark's `videoMs` was derived. */
export type AnchorMethod = 'direct' | 'calibrated' | 'none';

export interface MarkAnchorInfo {
  method: AnchorMethod;
  /** `GetRecordStatus.outputDuration` from the anchoring status call, ms. */
  outputDurationMs?: number;
  /** Round-trip time of that status call, ms. */
  rttMs?: number;
}

export interface Mark {
  id: string;
  /** 1-based press order within the session. */
  seq: number;
  kind: MarkKind;
  /** Semantic label, e.g. "mark", "issue". Feeds chapter names and the report. */
  label: string;
  /** Accelerator that fired, e.g. "F8" (manual marks only). */
  hotkey?: string;
  /** Monotonic stamp taken first thing in the hotkey handler. */
  monoMs: number;
  /** Anchored position in the recording; null if anchoring failed entirely. */
  videoMs: number | null;
  /** In-game time when the target is an instrumented game (§3.5); else null. */
  gameTimeMs?: number | null;
  anchor?: MarkAnchorInfo;
}

export type SessionEventType =
  | 'record-started'
  | 'record-paused'
  | 'record-resumed'
  | 'record-stopped'
  | 'obs-disconnected'
  | 'obs-reconnected'
  | 'telemetry-connected'
  | 'telemetry-lost'
  | 'chapter-create-failed';

export interface SessionEvent {
  type: SessionEventType;
  monoMs: number;
  videoMs?: number | null;
  detail?: string;
}

/** Sparse samples of game telemetry, for timeline reconstruction. */
export interface TelemetrySample {
  monoMs: number;
  gameTimeMs: number;
  paused: boolean;
}

export interface ObsInfo {
  obsVersion?: string;
  obsWebSocketVersion?: string;
  /** Profile parameter AdvOut/RecFormat2 at record time (want "hybrid_mp4"). */
  recFormat?: string;
  /** Audio track bitmask AdvOut/RecTracks at record time (want mic on track 2). */
  recTracks?: string;
}

export interface SessionInfo {
  /** Directory-safe id, e.g. "2026-08-22_1530_dungeon-sweeper". */
  id: string;
  title: string;
  startedAtWall: string; // ISO 8601
  endedAtWall?: string;
  captureTarget: CaptureTarget;
  /** Absolute path OBS reported from StopRecord.outputPath. */
  recordingFile?: string;
  obs?: ObsInfo;
}

export interface SessionSidecar {
  schemaVersion: typeof SIDECAR_SCHEMA_VERSION;
  session: SessionInfo;
  marks: Mark[];
  events: SessionEvent[];
  telemetry: TelemetrySample[];
}

/** One line of `session.journal.jsonl` — append-only crash-safety log. */
export type JournalLine =
  | { t: 'session'; data: SessionInfo }
  | { t: 'mark'; data: Mark }
  | { t: 'event'; data: SessionEvent }
  | { t: 'telemetry'; data: TelemetrySample }
  | { t: 'end'; data: { endedAtWall: string; recordingFile?: string } };

export function emptySidecar(session: SessionInfo): SessionSidecar {
  return {
    schemaVersion: SIDECAR_SCHEMA_VERSION,
    session,
    marks: [],
    events: [],
    telemetry: [],
  };
}

/** Rebuild a sidecar from journal lines (order-preserving; last wins for session info). */
export function sidecarFromJournal(lines: JournalLine[]): SessionSidecar | null {
  let session: SessionInfo | null = null;
  const marks: Mark[] = [];
  const events: SessionEvent[] = [];
  const telemetry: TelemetrySample[] = [];
  for (const line of lines) {
    switch (line.t) {
      case 'session':
        session = line.data;
        break;
      case 'mark':
        marks.push(line.data);
        break;
      case 'event':
        events.push(line.data);
        break;
      case 'telemetry':
        telemetry.push(line.data);
        break;
      case 'end': {
        const current = session;
        if (current) {
          const updated: SessionInfo = { ...current, endedAtWall: line.data.endedAtWall };
          if (line.data.recordingFile) updated.recordingFile = line.data.recordingFile;
          session = updated;
        }
        break;
      }
    }
  }
  if (!session) return null;
  return { schemaVersion: SIDECAR_SCHEMA_VERSION, session, marks, events, telemetry };
}

File: pipeline/playtest_pipeline/session.py
"""Load the session sidecar (source of truth for marks).

Mirrors packages/shared/src/session.ts and schema/session.schema.json.
If session.json is missing or corrupt (recorder crash), rebuilds it from the
append-only session.journal.jsonl.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

SIDECAR_FILENAME = "session.json"
JOURNAL_FILENAME = "session.journal.jsonl"


@dataclass
class Mark:
    id: str
    seq: int
    kind: str  # manual | vad | auto
    label: str
    mono_ms: float
    video_ms: int | None
    game_time_ms: int | None = None
    hotkey: str | None = None

    @staticmethod
    def from_json(d: dict) -> "Mark":
        return Mark(
            id=d["id"],
            seq=d["seq"],
            kind=d.get("kind", "manual"),
            label=d.get("label", "mark"),
            mono_ms=d.get("monoMs", 0),
            video_ms=d.get("videoMs"),
            game_time_ms=d.get("gameTimeMs"),
            hotkey=d.get("hotkey"),
        )


@dataclass
class Session:
    id: str
    title: str
    started_at_wall: str
    capture_target_kind: str
    capture_target_name: str | None
    recording_file: str | None
    marks: list[Mark] = field(default_factory=list)
    raw: dict = field(default_factory=dict)

    @staticmethod
    def from_sidecar(data: dict) -> "Session":
        if data.get("schemaVersion") != 1:
            raise ValueError(f"Unsupported sidecar schemaVersion: {data.get('schemaVersion')!r}")
        info = data["session"]
        target = info.get("captureTarget", {})
        return Session(
            id=info["id"],
            title=info.get("title", info["id"]),
            started_at_wall=info.get("startedAtWall", ""),
            capture_target_kind=target.get("kind", "other"),
            capture_target_name=target.get("name"),
            recording_file=info.get("recordingFile"),
            marks=[Mark.from_json(m) for m in data.get("marks", [])],
            raw=data,
        )


def _sidecar_from_journal(journal_path: Path) -> dict | None:
    """Rebuild a sidecar dict from journal lines (mirrors sidecarFromJournal in TS)."""
    session_info: dict | None = None
    marks: list[dict] = []
    events: list[dict] = []
    telemetry: list[dict] = []
    for raw_line in journal_path.read_text(encoding="utf-8").splitlines():
        raw_line = raw_line.strip()
        if not raw_line:
            continue
        try:
            line = json.loads(raw_line)
        except json.JSONDecodeError:
            continue  # torn tail write from a crash — skip
        t, data = line.get("t"), line.get("data")
        if t == "session":
            session_info = data
        elif t == "mark":
            marks.append(data)
        elif t == "event":
            events.append(data)
        elif t == "telemetry":
            telemetry.append(data)
        elif t == "end" and session_info is not None:
            session_info["endedAtWall"] = data.get("endedAtWall")
            if data.get("recordingFile"):
                session_info["recordingFile"] = data["recordingFile"]
    if session_info is None:
        return None
    return {
        "schemaVersion": 1,
        "session": session_info,
        "marks": marks,
        "events": events,
        "telemetry": telemetry,
    }


def load_session(session_dir: Path) -> Session:
    sidecar_path = session_dir / SIDECAR_FILENAME
    journal_path = session_dir / JOURNAL_FILENAME
    data: dict | None = None
    if sidecar_path.exists():
        try:
            data = json.loads(sidecar_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            data = None
    if data is None and journal_path.exists():
        print(f"  session.json missing/corrupt — rebuilding from {JOURNAL_FILENAME}")
        data = _sidecar_from_journal(journal_path)
    if data is None:
        raise FileNotFoundError(
            f"No readable {SIDECAR_FILENAME} or {JOURNAL_FILENAME} in {session_dir}"
        )
    return Session.from_sidecar(data)


def resolve_recording(session: Session, session_dir: Path, override: str | None) -> Path:
    """Find the OBS recording: CLI override > sidecar path > any video in session dir."""
    if override:
        p = Path(override)
        if p.exists():
            return p
        raise FileNotFoundError(f"--recording not found: {override}")
    if session.recording_file:
        p = Path(session.recording_file)
        if not p.is_absolute():
            p = session_dir / p
        if p.exists():
            return p
        print(f"  sidecar recording path missing on disk: {p}")
    candidates = [
        f for ext in ("*.mp4", "*.mkv", "*.mov") for f in session_dir.glob(ext)
    ]
    if candidates:
        best = max(candidates, key=lambda f: f.stat().st_size)
        print(f"  using recording found in session dir: {best.name}")
        return best
    raise FileNotFoundError(
        "Recording not found. Pass --recording PATH (the OBS output file for this session)."
    )
